### PR TITLE
Stringify visitor cosmetic improvements and debug feature

### PR DESCRIPTION
**Pull request by vladkluev**
This is pulling out the StringifyVisitor changes from #2037 and putting them into their own PR
## This PR does the following

* Adds the option to turn off pretty printing expressions
* Adds an `initial_indent` to the `StringifyVisitor` class (for use with other formatters)
* Wraps aliased subscriptable references in parentheses
* Removes an unnecessary newline at the beginning of formatted expression

### CODE REVIEW

File: snuba/query/expressions.py [hunk 1]
```diff
@@ -147,62 +165,92 @@ def visit_literal(self, exp: Literal) -> str:
...
+        return [f"{subscripted_column_str}{self._get_alias_str(exp)}"]
 
-    def visit_function_call(self, exp: FunctionCall) -> str:
+    def visit_function_call(self, exp: FunctionCall) -> List[str]:
         self.__level += 1
-        # every param will have a newline in front of it so no need to put any space
-        # after the comma
-        param_str = ",".join([param.accept(self) for param in exp.parameters])
+        param_strs = [
+            f"{param},"
```
Comment (evanh) [inline snuba/query/expressions.py:202]: Doesn't this inject extra commas? What if I have a function that contains another function as a parameter?

Like say `divide(count(), 2)`. `param.accept(self)` on the first parameter will return `["count(", ")"]`. You will then end up with a param_strs of `["count(,", "),", "2"]`. Unless I'm missing something?

**Review by evanh — COMMENTED** [review]
There is a code smell here where in a few of the visitor functions (subscriptable, function etc.) you are assuming that each response only returns a single string. For example in the visit_subscriptable function you assume the `exp.key.accept` and `exp.column.accept` return a list with a single value.

I think if we are going to have this return a list of strings, each function should return a list of tokens, and then there will have to be separate logic to determine if those tokens should be joined directly (`"".join(tokens)`) or joined with newlines.

File: snuba/query/expressions.py [hunk 2]
```diff
@@ -3,13 +3,23 @@
...
+from typing import Callable, Generic, Iterator, List, Optional, Tuple, TypeVar, Union
+
+from snuba import settings
 
 TVisited = TypeVar("TVisited")
+# daaclasses have their own built in repr, we override it
+# under usual circumstances to get more readable expressions.
 
+# The dataclass repr is created by the @dataclass decorator at
+# class definition time, therefore we have to know up front whether we
```
Comment (fpacifici) [inline snuba/query/expressions.py:16]: I think `upfront` is spelled as one word.

File: snuba/query/expressions.py [hunk 3]
```diff
@@ -147,62 +165,92 @@ def visit_literal(self, exp: Literal) -> str:
...
 
-    def visit_function_call(self, exp: FunctionCall) -> str:
+    def visit_function_call(self, exp: FunctionCall) -> List[str]:
         self.__level += 1
-        # every param will have a newline in front of it so no need to put any space
-        # after the comma
-        param_str = ",".join([param.accept(self) for param in exp.parameters])
+        param_strs = [
+            f"{param},"
+            for param in chain(*(param.accept(self) for param in exp.parameters))
```
Comment (fpacifici) [inline snuba/query/expressions.py:203]:
```
param_strs = [
   f"{param.accept(self}," for param in exp.parameters
]
```
but I think your issue will be that you are turning all parameters to strings while they can be function calls themselves. Shouldn't they be added as lists to the current list ?

**Comment by vladkluev** [comment]
> I think if we are going to have this return a list of strings, each function should return a list of tokens, and then there will have to be separate logic to determine if those tokens should be joined directly ("".join(tokens)) or joined with newlines.

I agree with this in principle but this once again makes things more complicated. So given this PR and the extra complexity it introduces. I'll ask again if the juice is worth the squeeze to make this interface "nicer" (as requested in the original PR). I would lean on no. I would be open to making a function like:

```
def format_as_str_list(expr: Expression, intial_indent: int = 0) -> List[str]:
    return expr.accept(StringifyVisitor(intial_indent)).split("\n")
```

in this module and using that in the query formatter instead of the StringifyVisitor directly. That will keep the implementation details of the StringifyVisitor in this module and the tracing formatter doesn't have to care about the formatting details. Sound good?

**Comment by vladkluev** [comment]
☝️ @fpacifici @evanh

**Comment by vladkluev** [comment]
Actually I take what I said back about the function that returns a list of strings. Doing that would make the query formatter code annoying as well for exactly the same reason that @evanh pointed out: not all lines are created equal, we would need to output tokens and then have the token reader decide how to combine them. Overall I don't see a benefit in introducing that extra complexity for something that is a debug feature. Happy to be convinced otherwise though.

**Comment by fpacifici** [comment]
I think I know what our issues is. We are trying to squeeze a hierarchical representation (expression or query) into a list of strings. a List then we find ourselves adding hacks not to loose the hierarchical nature of the data structure.
I remember dealing with this problem when I wrote the formatter for the Clickhosue query.
https://github.com/getsentry/snuba/blob/master/snuba/clickhouse/formatter/query.py#L30

I had to introduce an intermediate representation to be able to format the query either as a string or as something good for tracing and debugging
https://github.com/getsentry/snuba/blob/master/snuba/clickhouse/formatter/nodes.py

That generated a hierarchical structure that was agnostic to the type of nodes it was representing but that was capable of being serialized in multiple ways. 

Not sure with all these back and forth if it is worth going that way for expression and query. Up to you.
I would be ok with the original string based representation. If you want to go that way I will accept the query one. We can come back to this again.

**Comment by evanh** [comment]
I'm also fine with the original implementation. This is becoming way too complicated.

**Review by evanh — APPROVED** [review]

**Pull request merged**